### PR TITLE
feat: invalidate moderation and unblock user

### PR DESCRIPTION
**Pull request by mantagen**
## Description

- ~adds `/api/admin/...` route to invalidate a moderation~
- adds trpc admin procedure and router
- adds admin page for chat with moderations on, with button to invalidate a moderation 
- adds methods to safety violations class to revert a safety violation
- adds method to moderations class to invalidate a moderation
- updates moderation slack notification with a link to the 'invalidate moderation' link
- adds relevant migration (which has been run against staging)

**ps: excuse the styles, it's internal right?**

## Issue(s)

https://www.notion.so/oaknationalacademy/Be-able-to-de-block-a-lesson-that-was-deemed-toxic-cf42d50e12b64d649bd95207aaac5baf?pvs=4

## How to test

1. Go to the app
2. make a lesson and get a toxic moderation e.g. using mod:tox
3. go to #ai-dev-user-events in slack
4. click "Admin chat"
5. click invalidate next to your moderation
6. you should have the moderation invalidated and the chat be available again

**Note: you can also get yourself banned and test that it unbans you **

## Screenshots

image

### CODE REVIEW

File: apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      },
+    );
+  }
+
+  return new Response("Moderation invalidated and safety violation removed", {
+    status: 200,
+  });
+}
+
+export const GET = handler;
```
Comment (mantagen) [inline apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts:51]: Traditionally more of a POST kinda thing if strictly REST, but as we want people to be able to just click it from a slack message, GET made sense

Comment (stefl) [reply apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts:51]: Is there any chance a slack message ends up being sent as an email to anyone? If so, the mail client might fetch the URL to see if it is valid / anti-spam

Comment (stefl) [reply apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts:51]: Might be better to go to a page that has a form on it if so

File: packages/core/src/utils/slack.ts [hunk 2]
```diff
@@ -52,26 +52,49 @@ export function userIdBlock(userId: string): SectionBlock {
...
-export function userButtonsBlock(userId: string): ActionsBlock {
-  return {
-    type: "actions",
-    elements: [
-      {
-        type: "button",
-        text: {
-          type: "plain_text",
-          text: "Posthog User",
+export function actionsBlock({
```
Comment (mantagen) [inline packages/core/src/utils/slack.ts:55]: Initially i just wanted to add another block ... but that was failing silently and it turns out you're only allowed one actions block per message

File: apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  invalidatedBy: string;
+}) {
+  // invalidate toxic moderation
+  console.log("Invalidating moderation", { moderationId, invalidatedBy });
+  const moderations = new Moderations(prisma);
+  await moderations.invalidateModeration({
+    moderationId,
+    invalidatedBy,
+  });
+  console.log("Moderation invalidated");
```
Comment (stefl) [inline apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts:76]: Do we need?

File: packages/core/src/functions/slack/notifyModeration.ts [hunk 4]
```diff
@@ -42,7 +43,7 @@ export const notifyModeration = inngest.createFunction(
             fields: [
               {
                 type: "mrkdwn",
-                text: `*Chat*: <https://labs.thenational.academy/aila/${args.chatId}|aila/${args.chatId}>`,
+                text: `*Chat*: <https://${process.env.VERCEL_URL}/aila/${args.chatId}|aila/${args.chatId}>`,
```
Comment (stefl) [inline packages/core/src/functions/slack/notifyModeration.ts:46]: This won't be the labs dot domain, it'll be the ...vercel.app - will that cause a clerk issue?

File: packages/core/src/models/safetyViolations.ts [hunk 5]
```diff
@@ -123,4 +123,52 @@ export class SafetyViolations {
...
+
+    if (user.banned) {
+      this.logger.info(`Unbanning user ${userId}`);
+      await clerkClient.users.unbanUser(userId);
+
+      posthogAiBetaServerClient.capture({
+        distinctId: userId,
+        event: "User Unbanned",
+      });
+      posthogAiBetaServerClient.identify({
```
Comment (stefl) [inline packages/core/src/models/safetyViolations.ts:168]: Does this need to await .flush()

**Review by stefl — COMMENTED** [review]
Not sure that doing this with GET is going to be safe because Slack might have unknown behaviour?

File: apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts (hunk 3 above)
Comment (mantagen) [reply apps/nextjs/src/app/api/admin/moderation/[moderationId]/invalidate/route.ts]: Removed

File: packages/core/src/models/safetyViolations.ts (hunk 5 above)
Comment (mantagen) [reply packages/core/src/models/safetyViolations.ts:168]: Yep good spot

File: packages/core/src/functions/slack/notifyModeration.ts (hunk 4 above)
Comment (mantagen) [reply packages/core/src/functions/slack/notifyModeration.ts:46]: Good thinking, have amended

**Review by codeincontext — APPROVED** [review]

**Pull request merged**